### PR TITLE
Data.Dict: Add .from_list()

### PR DESCRIPTION
**Pull request by thinca**
`Data.Dict.from_array()` can make a dictionary from an array like `[key, value, key, value, ...]` or `[[key, value], [key, value], ...]`.

### CODE REVIEW

**Comment by ujihisa** [comment]
:+1::sushi:

File: doc/vital/Data/Dict.txt [hunk 1]
```diff
@@ -21,6 +21,13 @@ INTERFACE				*Vital.Data.Dict-interface*
 ------------------------------------------------------------------------------
 FUNCTIONS				*Vital.Data.Dict-functions*
 
+from_array({key-value-array})		*Vital.Data.Dict.from_array()*
+	Makes a dictionary from {key-value-array}.
+	{key-vakue-array} is one of following forms: >
```
Comment (tyru) [inline doc/vital/Data/Dict.txt]: s/vakue/value/

File: test/Data/Dict.vimspec [hunk 2]
```diff
@@ -3,6 +3,29 @@ Describe Data.Dict
...
+    End
+
+    Context with an array like [key, value, [key, value]]
+      It makes a new dictionary
+        let array = ['one', 1, ['two', 2], 'three', [3]]
+        Assert Equals(D.from_array(array), {'one': 1, 'two': 2, 'three': [3]})
+      End
+    End
+  End
+
```
Comment (tyru) [inline test/Data/Dict.vimspec]: Should we add error case?

File: autoload/vital/__vital__/Data/Dict.vim [hunk 3]
```diff
@@ -3,12 +3,39 @@
...
+function! s:_vital_depends() abort
+  return ['Vim.Type']
+endfunction
+
+
+function! s:from_array(array) abort
+  let dict = {}
+  let i = 0
+  let len = len(a:array)
+  while i < len
```
Comment (mattn) [inline autoload/vital/__vital__/Data/Dict.vim]: raise exception when the len is odd number?

File: autoload/vital/__vital__/Data/Dict.vim [hunk 4]
```diff
@@ -3,12 +3,39 @@
...
+    if type(a:array[i]) == s:t.string
+      let dict[a:array[i]] = a:array[i + 1]
```
Comment (mattn) [inline autoload/vital/__vital__/Data/Dict.vim]: why don't you check the type of key?

File: autoload/vital/__vital__/Data/Dict.vim [hunk 5]
```diff
@@ -3,12 +3,39 @@
...
+      let i += 2
+    else
+      " Assumes the value is array
+      let key_value = a:array[i]
+      let dict[key_value[0]] = key_value[1]
```
Comment (mattn) [inline autoload/vital/__vital__/Data/Dict.vim]: ditto

File: autoload/vital/__vital__/Data/Dict.vim (hunk 3 above)
Comment (tyru) [reply autoload/vital/__vital__/Data/Dict.vim]: This function takes a array like `['one', 1, ['two', 2], 'three', [3]]` (see the test case).

File: autoload/vital/__vital__/Data/Dict.vim (hunk 4 above)
Comment (tyru) [reply autoload/vital/__vital__/Data/Dict.vim]: ?
previous line checks the type, doesn't it?

Comment (mattn) [reply autoload/vital/__vital__/Data/Dict.vim]: ah, sorry. ignore this.

File: test/Data/Dict.vimspec (hunk 2 above)
Comment (thinca) [reply test/Data/Dict.vimspec]: I added some error handlings and tests for these.

**Comment by thinca** [comment]
The term in Vim is `list` not `array` so I renamed these :bow:

File: doc/vital/Data/Dict.txt [hunk 6]
```diff
@@ -21,6 +21,13 @@ INTERFACE				*Vital.Data.Dict-interface*
 ------------------------------------------------------------------------------
 FUNCTIONS				*Vital.Data.Dict-functions*
 
+from_list({key-value-list})		*Vital.Data.Dict.from_list()*
+	Makes a dictionary from {key-value-list}.
+	{key-value-list} is one of following forms: >
+	[key, value, key, value]
+<	or >
+	[[key, value], [key, value]]
```
Comment (tyru) [inline doc/vital/Data/Dict.txt]: * Add the description of the 3rd case

```
or mixed:
[key, value, [key, value]]
```

* Add "`Please note that key must be string`"

File: autoload/vital/__vital__/Data/Dict.vim [hunk 7]
```diff
@@ -3,12 +3,44 @@
...
+  let dict = {}
+  let i = 0
+  let len = len(a:list)
+  while i < len
+    if type(a:list[i]) == s:t.list
+      let key_value = a:list[i]
+      if len(key_value) != 2
+        throw 'vital: Data.Dict: Invalid key-value pair index at ' . i
+      endif
+      let dict[key_value[0]] = key_value[1]
```
Comment (tyru) [inline autoload/vital/__vital__/Data/Dict.vim]: As @mattn said [here](https://github.com/vim-jp/vital.vim/pull/550#discussion_r150187778), we should add the type check of key (`a:list[i]`) and raise an error

File: autoload/vital/__vital__/Data/Dict.vim [hunk 8]
```diff
@@ -3,12 +3,44 @@
...
+      let i += 1
+    else
+      if len <= i + 1
+        throw 'vital: Data.Dict: Invalid key-value pair index at ' . i
+      endif
+      let dict[a:list[i]] = a:list[i + 1]
```
Comment (tyru) [inline autoload/vital/__vital__/Data/Dict.vim]: ditto (add the type check of key)

File: test/Data/Dict.vimspec [hunk 9]
```diff
@@ -3,6 +3,43 @@ Describe Data.Dict
...
+    End
+
+    Context when there is no value corresponding to the key
+      It throws an error
+        let list = ['key1', 'value1', 'key2']
+        Throws /^vital: Data.Dict:/ D.from_list(list)
+      End
+    End
+  End
+
```
Comment (tyru) [inline test/Data/Dict.vimspec]: Add the test case(s) of wrong type of key (key must be string, or number? (not float))

These cases are clearly error:

```vim
let list = [1.1, 'value1']
let list = [{}, 'value1']
```

However, I don't think all type cases should be covered.

**Comment by tyru** [comment]
I think the type of key which should be allowed are:

* string
* number (not float)

**Comment by thinca** [comment]
I'll squash when reviews finished.

**Comment by tyru** [comment]
Seems good except https://github.com/vim-jp/vital.vim/pull/550/files#r150204801

File: doc/vital/Data/Dict.txt (hunk 6 above)
Comment (thinca) [reply doc/vital/Data/Dict.txt]: > Add the description of the 3rd case

This is extra form.  This function works like that, however I do not want to explicitly describe that in the document.

> Add "Please note that key must be string"

Hmm, this is basic of Vim script.  If need this sentence in here, we should write same sentence to description of many other functions.

Comment (tyru) [reply doc/vital/Data/Dict.txt]: Okay.
My comment was just suggestions.
so it's OK to determine that you don't write them in document.

**Comment by thinca** [comment]
Thanks all!

**Pull request merged**